new-log-viewer/src/services/decoders/JsonlDecoder/index.ts

```diff
@@ -0,0 +1,244 @@
+import {
+    INVALID_TIMESTAMP_VALUE,
+    JsonLogEvent,
+    LOG_LEVEL,
+    LogLevelFilter,
+} from "../../../typings/logs";
+import LogbackFormatter from "../../formatters/LogbackFormatter";
+import {
+    convertToDayjsTimestamp,
+    convertToLogLevelValue,
+    isJsonObject,
+} from "./utils";
+
+
+/**
+ * A decoder for JSONL (JSON lines) files that contain log events. See `JsonlDecoderOptionsType` for
+ * properties that are specific to log events (compared to generic JSON records).
+ */
+class JsonlDecoder implements Decoder {
+    static #textDecoder = new TextDecoder();
+
+    #dataArray: Nullable<Uint8Array>;
+
+    #logLevelKey: string;
+
+    #timestampKey: string;
+
+    #logEvents: JsonLogEvent[] = [];
+
+    #filteredLogEventMap: FilteredLogEventMap = null;
+
+    #invalidLogEventIdxToRawLine: Map<number, string> = new Map();
+
+    #formatter: Formatter;
+
+    /**
+     * @param dataArray
+     * @param decoderOptions
+     */
+    constructor (dataArray: Uint8Array, decoderOptions: JsonlDecoderOptionsType) {
+        this.#dataArray = dataArray;
+        this.#logLevelKey = decoderOptions.logLevelKey;
+        this.#timestampKey = decoderOptions.timestampKey;
+        this.#formatter = new LogbackFormatter({formatString: decoderOptions.formatString});
+    }
+
+    getEstimatedNumEvents (): number {
+        return this.#logEvents.length;
+    }
+
+    getFilteredLogEventMap (): FilteredLogEventMap {
+        return this.#filteredLogEventMap;
+    }
+
+    setLogLevelFilter (logLevelFilter: LogLevelFilter): boolean {
+        this.#filterLogEvents(logLevelFilter);
+
+        return true;
+    }
+
+    build (): LogEventCount {
+        this.#deserialize();
+
+        const numInvalidEvents = Array.from(this.#invalidLogEventIdxToRawLine.keys()).length;
+
+        return {
+            numValidEvents: this.#logEvents.length - numInvalidEvents,
+            numInvalidEvents: numInvalidEvents,
+        };
+    }
+
+    setFormatterOptions (options: JsonlDecoderOptionsType): boolean {
+        this.#formatter = new LogbackFormatter({formatString: options.formatString});
+
+        return true;
+    }
+
+    decodeRange (
+        beginIdx: number,
+        endIdx: number,
+        useFilter: boolean,
+    ): Nullable<DecodeResultType[]> {
+        if (useFilter && null === this.#filteredLogEventMap) {
+            return null;
+        }
+
+        const length: number = (useFilter && null !== this.#filteredLogEventMap) ?
+            this.#filteredLogEventMap.length :
+            this.#logEvents.length;
+
+        if (0 > beginIdx || length < endIdx) {
+            return null;
+        }
+
+        const results: DecodeResultType[] = [];
+        for (let i = beginIdx; i < endIdx; i++) {
+            // Explicit cast since typescript thinks `#filteredLogEventMap[i]`
+            // can be undefined, but it shouldn't be since we performed a bounds check at the
+            // beginning of the method.
```

Comment:
```suggestion
            // Explicit cast since typescript thinks `#filteredLogEventMap[i]` can be undefined, but
            // it shouldn't be since we performed a bounds check at the beginning of the method.
```